ipld/src/selector/mod.rs

```diff
@@ -236,9 +235,126 @@ impl Selector {
+                        PathSegment::Int(i) => fields.remove(&format!("{}", i)),
+                    }
+                }
+                _ => None,
+            },
+            ExploreIndex { index, next } => match ipld {
+                Ipld::List(l) => {
+                    let i = p.to_index()?;
+                    if i != index || i >= l.len() {
+                        None
+                    } else {
+                        // Path segment matches selector index
+                        Some(*next)
+                    }
+                }
+                _ => None,
+            },
+            ExploreRange { start, end, next } => {
+                match ipld {
+                    Ipld::List(l) => {
+                        let i = p.to_index()?;
+                        // Check to make sure index is within list bounds
+                        if i < start || i >= end || i >= l.len() {
+                            None
+                        } else {
+                            // Path segment is within the selector range
+                            Some(*next)
+                        }
+                    }
+                    _ => None,
+                }
+            }
+            ExploreRecursive {
+                current,
+                sequence,
+                limit,
+                stop_at,
+            } => {
+                let next = match current {
+                    Some(s) => s.explore(ipld, p)?,
+                    None => sequence.clone().explore(ipld, p)?,
+                };
+
+                if !has_recursive_edge(&next) {
+                    return Some(ExploreRecursive {
+                        sequence,
+                        current: Some(next.into()),
+                        limit,
+                        stop_at,
+                    });
+                }
+                match limit {
+                    RecursionLimit::Depth(depth) => {
+                        if depth < 2 {
+                            // Replaces recursive edge with None on last iteration
+                            // TODO revisit, shouldn't need to replace, would be better to just
+                            // return none when edge is hit on final depth
+                            return replace_recursive_edge(next, None).map(|s| *s);
+                        }
+                        Some(ExploreRecursive {
+                            current: replace_recursive_edge(next, Some(sequence.clone())),
+                            sequence,
+                            limit: RecursionLimit::Depth(depth - 1),
+                            stop_at,
+                        })
+                    }
+                    RecursionLimit::None => Some(ExploreRecursive {
+                        current: replace_recursive_edge(next, Some(sequence.clone())),
+                        sequence,
+                        limit,
+                        stop_at,
+                    }),
+                }
```

Comment:
Some duplication going on here with only `limit` being different, I suggest
```rust
RecursionLimit::Depth(depth) if depth < 2 => ...,
_ => ...,
```
That would probably work best by giving the `RecursionLimit` type something like a `decr()` method

Reply:
what you suggest would break the logic, since Depth and None variant's still need to be differentiated, I will add the decrement method and just return outside of the match to avoid the dedup though, that's a good idea.

Reply:
decided to go with a SubAssign<u64> impl, seems cleanest way

Reply:
The `decr` I suggested would just do nothing in case of `RecursionLimit::None`, so it would have worked in pretty much the same way 🙂

Reply:
Yeah the decr part of the suggestion was valid for sure, switching to subassign was just to be able to do `limit -=1;` instead of `limit.decr()` (only an opinionated change), I was just mentioning that the code block you put wrt breaking the logic

Reply:
Actually, in your current code structure you could just do `*depth -= 1` directly (with a `ref mut` binding), no need to implement it on `RecursionLimit`

Reply:
very true! Idea was for future use, but I guess we can assume this decrement will never be needed outside of a match

Reply:
Idk I kinda think this is a bit less readable/usable actually though, thoughts?
```rust
                if let RecursionLimit::Depth(depth) = &mut limit {
                    if *depth < 2 {
                        // Replaces recursive edge with None on last iteration
                        // TODO revisit, shouldn't need to replace, would be better to just
                        // return none when edge is hit on final depth
                        return replace_recursive_edge(next, None);
                    }
                    *depth -= 1;
                }
```

Reply:
Yeah, I have to agree, the mutation makes this less readable... Here's a tiny adaptation to consider:
```rust
let new_limit = match limit {
    RecursionLimit::Depth(depth) => {
        if depth < 2 {
            // Replaces recursive edge with None on last iteration
            // TODO revisit, shouldn't need to replace, would be better to just
            // return none when edge is hit on final depth
            return replace_recursive_edge(next, None);
        }
        RecursionLimit::Depth(depth - 1)
    }
    RecursionLimit::None => RecursionLimit::None,
};
```
Takes slightly less mental juggling to read this, I think.

Reply:
This just adds complexity to the way it is rn. I think, although useful, have gone too deep into this minor detail which has a good change of being refactored later anyway

Reply:
It's indeed pretty inconsequential 🙂 Though it definitely does not add complexity compared to mutability + the `SubAssign` conformance

Reply:
Your original version was very simple though, that might be the best one for now